pkg/build/webhook/webhook.go

```diff
@@ -47,53 +52,86 @@ func GitRefMatches(eventRef, configRef string, buildSource *buildapi.BuildSource
 	return configRef == eventRef
 }
 
-// FindTriggerPolicy retrieves the BuildTrigger of a given type from a build
-// configuration
-func FindTriggerPolicy(triggerType buildapi.BuildTriggerType, config *buildapi.BuildConfig) (buildTriggers []buildapi.BuildTriggerPolicy, err error) {
-	buildTriggers = buildapi.FindTriggerPolicy(triggerType, config)
-	if len(buildTriggers) == 0 {
-		err = ErrHookNotEnabled
-	}
-	return buildTriggers, err
+// NewWarning returns an StatusError object with a http.StatusOK (200) code.
+func NewWarning(message string) *kerrors.StatusError {
+	return &kerrors.StatusError{ErrStatus: metav1.Status{
+		Status:  metav1.StatusSuccess,
+		Code:    http.StatusOK,
+		Message: message,
+	}}
 }
 
-// ValidateWebHookSecret validates the provided secret against all currently
-// defined webhook secrets and if it is valid, returns its information.
-func ValidateWebHookSecret(webHookTriggers []buildapi.BuildTriggerPolicy, secret string) (*buildapi.WebHookTrigger, error) {
-	for _, trigger := range webHookTriggers {
-		switch trigger.Type {
-		case buildapi.GenericWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GenericWebHook.Secret), []byte(secret)) {
-				continue
-			}
-			return trigger.GenericWebHook, nil
-		case buildapi.GitHubWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GitHubWebHook.Secret), []byte(secret)) {
-				continue
+// CheckSecret tests the user provided secret against the secrets for the webhook triggers, if a match is found
+// then the corresponding webhook trigger is returned.
+func CheckSecret(namespace, userSecret string, triggers []*buildapi.WebHookTrigger, secretsClient kcoreclient.SecretsGetter) (*buildapi.WebHookTrigger, error) {
+	for i := range triggers {
+		secretRef := triggers[i].SecretReference
+		secret := triggers[i].Secret
+		if len(secret) > 0 {
+			if hmac.Equal([]byte(secret), []byte(userSecret)) {
+				return triggers[i], nil
 			}
-			return trigger.GitHubWebHook, nil
-
-		case buildapi.GitLabWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GitLabWebHook.Secret), []byte(secret)) {
-				continue
+		}
+		if secretRef != nil {
+			glog.V(4).Infof("Checking user secret against secret ref %s", secretRef.Name)
+			s, err := secretsClient.Secrets(namespace).Get(secretRef.Name, metav1.GetOptions{})
+			if err != nil && !kerrors.IsNotFound(err) {
+				return nil, err
 			}
-			return trigger.GitLabWebHook, nil
-
-		case buildapi.BitbucketWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.BitbucketWebHook.Secret), []byte(secret)) {
-				continue
+			if v, ok := s.Data[buildapi.WebHookSecretKey]; ok {
```

Comment:
ok is being checked by the if block.  if ok is false its going to fall through to return "errsecretmismatch".  I'll add an explicit TC for that scenario.


checking len(v) should be unnecessary since it's valid(if silly) to have your secret value be empty string (though i'm not even sure you could successfully pass emptystring in the webhook invocation)

Reply:
I guess I wanted to check if it was considered valid - before this PR is it possible to have an empty string secret?

Reply:
i don't think it was possible before since the secret value field had to be non-empty.

i'm still pretty sure it's not possible now either since the user provided secret comes from the webhook url path, so it wouldn't be possible to provide an "empty" value there, that I can think of, is it?

but i'm not overly concerned about someone creating a Secret with an actual WebHookSecretKey that has an empty value.

Reply:
ok - I was tempted to keep things symmetric but it's no biggie.